sdk/include/opentelemetry/sdk/trace/etw_data.h

```diff
@@ -0,0 +1,530 @@
+  };
+
+  // Sets an attribute on the Span. If the Span previously contained a mapping for
+  // the key, the old value is replaced.
+  virtual void SetAttribute(nostd::string_view key,
+                            const common::AttributeValue &value) noexcept override
+  {
+    // TODO: not implemented
+    UNREFERENCED_PARAMETER(key);
+    UNREFERENCED_PARAMETER(value);
+  };
+
+  /// <summary>
+  /// Update Span name
+  /// </summary>
+  /// <param name="name"></param>
+  /// <returns></returns>
+  virtual void UpdateName(nostd::string_view name) noexcept override
+  {
+    // TODO: not implemented
+    UNREFERENCED_PARAMETER(name);
+  }
+
+  /// <summary>
+  /// End Span
+  /// </summary>
+  /// <returns></returns>
+  virtual void End(const trace::EndSpanOptions & = {}) noexcept override
+  {
+    // TODO: signal this to owner
+  }
+
+  virtual trace::SpanContext GetContext() const noexcept
+  {
+    // TODO: not implemented
+    static trace::SpanContext nullContext{trace::SpanContext::GetInvalid()};
+    return nullContext;
+  }
+
+  /// <summary>
+  /// Check if Span is recording data
+  /// </summary>
+  /// <returns></returns>
+  virtual bool IsRecording() const noexcept override
+  {
+    // TODO: not implemented
+    return true;
+  }
+
+  virtual void SetToken(nostd::unique_ptr<context::Token> &&token) noexcept
+  {
+    // TODO: not implemented
+    UNREFERENCED_PARAMETER(token);
+  }
+
+  /// <summary>
+  /// Get Owner tracer of this Span
+  /// </summary>
+  /// <returns></returns>
+  trace::Tracer &tracer() const noexcept { return this->owner; };
+};
+
+/// <summary>
+/// stream::TraceProvider
+/// </summary>
+class TracerProvider : public trace::TracerProvider
+{
+public:
+  /// <summary>
+  /// Obtain a Tracer of given type (name) and supply extra argument arg2 to it.
+  /// </summary>
+  /// <param name="name">Tracer Type</param>
+  /// <param name="args">Tracer arguments</param>
+  /// <returns></returns>
+  virtual nostd::shared_ptr<trace::Tracer> GetTracer(nostd::string_view name,
+                                                     nostd::string_view args = "")
+  {
+    // TODO: describe possible args. Currently supported:
+    // "MSGPACK"        - use MessagePack
+    // "XML"            - use XML
+    // "ETW"            - use 'classic' Trace Logging Dynamic manifest ETW event
+    //
+#if defined(HAVE_NO_TLD) && defined(HAVE_MSGPACK)
+    ETWProvider::EventFormat evtFmt = ETWProvider::EventFormat::ETW_MSGPACK;
+#else
+    ETWProvider::EventFormat evtFmt = ETWProvider::EventFormat::ETW_MANIFEST;
+#endif
+
+#pragma warning(push)
+#pragma warning(disable : 4307) /* Integral constant overflow - OK while computing hash */
+    auto h = utils::hashCode(args.data());
+    switch (h)
+    {
+      case CONST_HASHCODE(MSGPACK):
+        // nobrk
+      case CONST_HASHCODE(MsgPack):
+        // nobrk
+      case CONST_HASHCODE(MessagePack):
+        evtFmt = ETWProvider::EventFormat::ETW_MSGPACK;
```

Comment:
how does a pseudo-random hash-value determines which event format to use ? Perhaps I am missing something ?

Reply:
It's not random.